extensions/route_map/search.rs

```diff
@@ -0,0 +1,105 @@
+use pyo3::prelude::*;
+use std::borrow::Cow;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+use crate::util::normalize_path;
+use crate::{get_base_components, util, wrappers, HandlerGroup, Node, RouteMap};
+
+pub(crate) fn find_insert_handler_group<'a>(
+    root: &'a mut Node,
+    plain_routes: &'a mut HashMap<String, HandlerGroup>,
+    path: &str,
+    path_parameters: &PyAny,
+    is_static: bool,
+    handler_group: HandlerGroup,
+) -> PyResult<&'a mut HandlerGroup> {
+    let py = path_parameters.py();
+    let path_parameters_vec: Vec<wrappers::PathParameter<'_>> = path_parameters.extract()?;
+    let handler_group: &mut HandlerGroup = if !path_parameters_vec.is_empty() || is_static {
+        let param_set = util::param_set(&path_parameters_vec)?;
+        let mut node = root;
+        for s in get_base_components(path) {
+            // Could we just assume a path segment that starts and ends
+            // with `{}` is a placeholder?
+            let is_placeholder =
+                s.starts_with('{') && s.ends_with('}') && param_set.contains(&s[1..s.len() - 1]);
+
+            node = if is_placeholder {
+                node.placeholder_child
+                    .get_or_insert_with(Box::<Node>::default)
+            } else {
+                node.children
+                    .entry(String::from(s))
+                    .or_insert_with(Node::default)
+            };
+        }
+        // Found where the handlers should be, get it, or add a new one
+        match node.handler_group {
+            Some(ref mut existing) => existing.merge(py, handler_group, path)?,
+            None => node.handler_group.insert(handler_group),
+        }
+    } else {
+        match plain_routes.entry(String::from(path)) {
+            Entry::Occupied(entry) => entry.into_mut().merge(py, handler_group, path)?,
+            Entry::Vacant(entry) => entry.insert(handler_group),
+        }
+    };
+    Ok(handler_group)
+}
+
+impl RouteMap {
+    pub(crate) fn find_handler_group<'a>(&'a self, full_path: &'a str) -> PyResult<FindResult<'a>> {
+        let mut path = normalize_path(full_path);
+        let mut param_values = Vec::new();
+        let mut node = &self.root;
+
+        let handler_group: &HandlerGroup =
+            if let Some(handler_group) = self.plain_routes.get(path.as_ref()) {
+                handler_group
+            } else {
+                for component in get_base_components(&path) {
```

Comment:
We could consider first storing the result of `get_base_components` in a variable and using its size as a "maximum" length of `param_values`, and optimistically allocate the vector size instead by calling `param_values.reserve()`. Definitely not a certainty that we would want to do this, but something to consider. The "slow" case is when there are a lot of nested nodes, and if many of them contain path parameters, there will most likely be more than one allocation. We could prevent this by always allocating with a capacity of `get_base_component`'s return value size, but it would add a constant single allocation for cases where it was never going to allocate (when placeholder_child never exists during traversal)

Reply:
I think the usual case will be to have a quite small number of placeholder components, so I think I'm okay with starting empty for now, we can benchmark if it will matter.